UserScrape/crawler.py

```diff
@@ -121,108 +135,145 @@ def login(self) -> CrawlResult:
+        return recos
+
+    def delete_last_video_from_history(self):
+        self.driver.get('https://www.youtube.com/feed/history')
+        # todo: maybe replace the full xpath because its ugly, but it works
+        last_video_delete_button = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.XPATH,
+                                            '/html/body/ytd-app/div/ytd-page-manager/ytd-browse/ytd-two-column-browse'
+                                            '-results-renderer/div[1]/ytd-section-list-renderer/div['
+                                            '2]/ytd-item-section-renderer[1]/div[3]/ytd-video-renderer/div['
+                                            '1]/div/div['
+                                            '1]/div/div/ytd-menu-renderer/div/ytd-button-renderer/a/yt-icon-button'
+                                            '/button'))
+        ).click()
+
+    def delete_history(self):
+        self.driver.get('https://www.youtube.com/feed/history')
+        history_delete_button = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.XPATH,
+                                            '/html/body/ytd-app/div/ytd-page-manager/ytd-browse/ytd-two-column-browse'
+                                            '-results-renderer/div['
+                                            '2]/ytd-browse-feed-actions-renderer/div/ytd-button-renderer['
+                                            '1]/a/paper-button/yt-formatted-string'))
+        ).click()
+
+        confirm_button = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.XPATH,
+                                            '/html/body/ytd-app/ytd-popup-container/paper-dialog/yt-confirm-dialog'
+                                            '-renderer/div[2]/div/yt-button-renderer['
+                                            '2]/a/paper-button/yt-formatted-string'))
+        ).click()
+
+    def _get_seconds(self, duration: str):
+        # helper function to correctly parse the time from the info bar
+        if len(duration) > 5:
+            duration_time = datetime.strptime(duration, "%H:%M:%S")
+        else:
+            duration_time = datetime.strptime(duration, "%M:%S")
+        return (duration_time-datetime(1900, 1, 1)).total_seconds()
+
+    def watch_video(self, videoId: str):
+        self.driver.get("https://www.youtube.com/watch?v=" + videoId)
+        # unfortunately YouTube orders the player in multiple different ways so we have to try them all unless there is a better way?
+        duration = WebDriverWait(self.driver, 3).until(
+                EC.presence_of_element_located((By.CLASS_NAME, 'ytp-time-duration'))).text
+        # start the video
+        duration = self._get_seconds(duration)
+        # to make sure that every video is watched long enough
+        # todo: this will be a problem if an ad is longer than the respective watch time?
```

Comment:
What do we do about ads? Do we skip them as soon as possible?
Do we try to detect them so that we can 'use' them?

Reply:
Can we detect what the title/company is for an ad? If we can we should capture it. Super interesting data.

Reply:
advertiser is now stored as a blob, ads are skipped